library/common/http/dispatcher.cc

```diff
@@ -49,75 +65,152 @@ void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool e
+  // false but remote_closed_ true. Because in that case the cleanup call is guarded by complete(),
+  // right?
+  http_dispatcher_.cleanup(direct_stream_.stream_handle_);
+}
+
+void Dispatcher::DirectStreamCallbacks::onCancel() {
+  // This call is guarded at the call-site @see Dispatcher::DirectStream::resetStream().
+  // Therefore, it is dispatched here without protection.
+  ENVOY_LOG(debug, "[S{}] dispatching to platform cancel stream", direct_stream_.stream_handle_);
+  bridge_callbacks_.on_cancel(bridge_callbacks_.context);
+
+  // FIXME: before merge. what about clean up? It all depends on if the reset we send down causes a
+  // reset coming up, which will not fire a callback, but will result in stream clean up.
+}
+
+Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle, Dispatcher& http_dispatcher)
+    : stream_handle_(stream_handle), parent_(http_dispatcher) {}
+
+void Dispatcher::DirectStream::resetStream(StreamResetReason) { callbacks_->onReset(); }
+
+void Dispatcher::DirectStream::closeLocal(bool end_stream) {
+  // TODO: potentially guard against double local closure.
+  local_closed_ = end_stream;
+
+  // No cleanup happens here because cleanup always happens on remote closure.
+  // see @Dispatcher::DirectStreamCallbacks::closeRemote.
+  // Remote closure is guaranteed to happen either through the golden path, or through an Envoy
+  // driven reset.
+}
+void Dispatcher::DirectStream::closeRemote(bool end_stream) {
+  remote_closed_ = end_stream;
+  // Even though we have potentially fired an on_complete callback to the bridge_callbacks,
+  // We will only cleanup if **both** the local and the remote have closed because otherwise we
+  // can be expecting a reset from envoy.
+  // See the comment for @see Dispatcher::DirectStreamCallbacks::closeRemote().
+  if ((complete())) {
+    ENVOY_LOG(debug, "[S{}] scheduling cleanup", stream_handle_);
+    parent_.cleanup(stream_handle_);
+  }
 }
 
-Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle,
-                                       AsyncClient::Stream& underlying_stream,
-                                       DirectStreamCallbacksPtr&& callbacks)
-    : stream_handle_(stream_handle), underlying_stream_(underlying_stream),
-      callbacks_(std::move(callbacks)) {}
-
-AsyncClient::StreamOptions
-Dispatcher::DirectStream::toNativeStreamOptions(envoy_stream_options stream_options) {
-  AsyncClient::StreamOptions native_stream_options;
-  native_stream_options.setBufferBodyForRetry(stream_options.buffer_body_for_retry);
-  return native_stream_options;
+bool Dispatcher::DirectStream::complete() { return local_closed_ && remote_closed_; }
+
+bool Dispatcher::DirectStream::dispatchable(bool close) {
+  // FIXME: before merge. discuss my thought that this state does not need to be atomic because it
+  // will all be accessed from within the context of Envoy's main thread. So all we need is an
```

Comment:
Cancellation is the case that won't be necessarily called from Envoy's main thread.

Reply:
This is related to the comment below where if cancellation state is handled in dispatched code then we are doing all the gating in Envoy's main thread.

Reply:
from offline: the contract with cancellation is that it is an interrupt. In other words, by the time this function returns no further callbacks associated with this stream will be surfaced. Hence we can't dispatch exchanging the closed state. It has async logic in it but the effect is sync.